x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/job/RollupIDGenerator.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.rollup.job;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefBuilder;
+import org.elasticsearch.common.Numbers;
+import org.elasticsearch.common.hash.MurmurHash3;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.zip.CRC32;
+
+/**
+ * The ID Generator creates a deterministic document ID to be used for rollup docs.
+ * It does this by accepting values (e.g. composite keys for the rollup bucket) and
+ * hashes those together in a deterministic manner.
+ *
+ * Once the ID has generated, the generator instance becomes locked and will not
+ * accept any more values.  This is a safety mechanism to prevent accidentally
+ * changing the ID at runtime.
+ */
+public abstract class RollupIDGenerator {
+    public abstract void add(Integer v);
+    public abstract void add(Long v);
+    public abstract void add(Double v);
+    public abstract void add(String v);
+    public abstract void addNull();
+    public abstract String getID();
+
+    private boolean generated = false;
```

Comment:
I haven't proved this to myself by writing a multithreaded driver, but, looking at the code, I am pretty certain that this class can't be used in a multithreaded env, because `check` and `setFlag` are not coordinated. if I am not wrong, we should probably either document this, or, potentially re-implement this class as a builder, where mutated methods return a modified instance? 🤷‍♂️

Reply:
I'll add a javadoc explaining the non-threadsafe nature.

This is used by the indexer which is single-threaded in Rollup (only a single thread-per-job is running the indexer).  We could make it an `AtomicBoolean` and make it thread safe, but given the current Rollup design it's overkill.  I don't see this being used outside of Rollup right now, so ++ to just documenting the limitation.